File: 프로그래머스/2/389480. 완전범죄/완전범죄.java
import java.lang.Math;
import java.util.Arrays;
class Solution {
    public int solution(int[][] info, int n, int m) {
        int answer = n;
        int[][] dp = new int[info.length+1][m+1];
        for(int i=0;i<info.length+1;i++){
            Arrays.fill(dp[i],n);
        }
        dp[0][0]=0;
        for(int i=1;i<info.length+1;i++){
            int a = info[i-1][0];
            int b = info[i-1][1];
            for(int j=0;j<m;j++){
                dp[i][j]=Math.min(dp[i][j],dp[i-1][j]+a);
                if(j+b<m){
                    dp[i][j+b]=Math.min(dp[i][j+b],dp[i-1][j]);
                }
            }
        }
        for(int i=0;i<m;i++){
            answer = Math.min(answer,dp[info.length][i]);
        }
        return answer==n?-1:answer;
    }
}

File: 프로그래머스/2/389480. 완전범죄/완전범죄.py
import copy
def solution(info, n, m):
    dp=[]
    for i in range(len(info)+1):
        line = [n]*m
        dp.append(copy.deepcopy(line))
    dp[0][0]=0
    for i in range(1,len(info)+1):
        a=info[i-1][0]
        b=info[i-1][1]
        for j in range(m):
            dp[i][j]=min([dp[i][j],dp[i-1][j]+a])
            if j+b<m:
                dp[i][j+b]=min([dp[i][j+b],dp[i-1][j]])
    answer = n
    for i in range(m):
        answer = min([answer,dp[len(info)][i]])
    return -1 if answer>=n else answer